### PR TITLE
Documentation update for dynamically loading MSAL configuration of MSAL Angular 2.x library using APP_INITIALIZER

### PR DESCRIPTION
**Pull request by vinusorout**
To load MSAL configurations dynamically we can use APP_INITIALIZER with Factory Providers. For a working example please check following sample https://github.com/vinusorout/msal-angular-dynamic-configuration-sample

### CODE REVIEW

**Comment by vinusorout** [comment]
Hello @jasonnutter , this is the first time I have created a PR here, and I couldn't figure out what is this Beachball Check and why it is failing. Could you please  let me know what to do to make this work!

**Comment by jasonnutter** [comment]
@vinusorout Navigating to the root of the repo, run `npm install` if you haven't already, then run `npm run beachball:change`, follow the prompts, and then push your updated branch to github.

**Comment by vinusorout** [comment]
Thanks for the quick response @jasonnutter , but after running this command I am getting following output:

WARN: There are untracked changes in your repository:
- lib/msal-angular/package-lock.json
- lib/msal-browser/package-lock.json
- lib/msal-common/package-lock.json
- lib/msal-core/package-lock.json
- lib/msal-node/package-lock.json
- lib/msal-react/package-lock.json
Changes in these files will not trigger a prompt for change descriptions
Checking for changes against "origin/dev"
fetching latest from remotes "origin/dev"
No change files are needed

 I am running this in, forked branch https://github.com/vinusorout/microsoft-authentication-library-for-js

**Comment by vinusorout** [comment]
Finally fixed the beachball check

**Comment by jo-arroyo** [comment]
@vinusorout I am unable to reproduce your instructions in our angular11-sample-app. Could you please ensure that no instructions are missing?

**Comment by vinusorout** [comment]
Hello @jo-arroyo I used these same files in a sample and working as expected, available here https://github.com/vinusorout/msal-angular-dynamic-configuration-sample

Could you please check this! Or let me know what exact error you are getting!

**Comment by vinusorout** [comment]
Hello @jo-arroyo , just fixed the configuration.json file path(assets/configuration.json) in app.module.ts file, it should work now, please check!

**Comment by vinusorout** [comment]
Hello @jo-arroyo have you got round to check this!

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 1]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+```typescript
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpBackend } from '@angular/common/http';
+import { map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ConfigService {
+
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:289]:
```suggestion

```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 2]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+  private settings: any;
+  private http: HttpClient;
+  constructor(private readonly httpHandler: HttpBackend) {
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:292]:
```suggestion

  constructor(private readonly httpHandler: HttpBackend) {
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 3]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+    this.http = new HttpClient(httpHandler);
+  }
+
+  init(endpoint: string): Promise<boolean> {
+    return new Promise<boolean>((resolve, reject) => {
+      this.http.get(endpoint).pipe(map(res => res))
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:298]:
```suggestion
      this.http.get(endpoint).pipe(map(result => result))
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 4]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+  getSettings(key?: string | Array<string>): any {
+    if (!key || (Array.isArray(key) && !key[0])) {
+      return this.settings;
+    }
+
+    if (!Array.isArray(key)) {
+      key = key.split('.');
+    }
+
+    let result = key.reduce((acc: any, current: string) => acc && acc[current], this.settings);
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:318]:
```suggestion
    let result = key.reduce((account: any, current: string) => account && account[current], this.settings);
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 5]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+    return result;
+  }
+}
+```
+
+`src/app/msal-config-dynamic.module.ts`
+```typescript
+import { InjectionToken, NgModule, APP_INITIALIZER } from '@angular/core';
+import { IPublicClientApplication, PublicClientApplication, 
+    InteractionType, BrowserCacheLocation, LogLevel } from '@azure/msal-browser';
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:329]:
```suggestion
    LogLevel } from '@azure/msal-browser';
```
No unused imports. Imports on same line if possible (we will be changing our other configuration methods to also do this)

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 6]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+import { MsalGuard, MsalInterceptor, MsalBroadcastService,
+     MsalInterceptorConfiguration, MsalModule, MsalService,
+      MSAL_GUARD_CONFIG, MSAL_INSTANCE, MSAL_INTERCEPTOR_CONFIG, 
+      MsalGuardConfiguration, MsalRedirectComponent } from '@azure/msal-angular';
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:333]:
```suggestion
      MsalGuardConfiguration } from '@azure/msal-angular';
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 7]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+        console.log('finished getting configurations dynamically.');
+    });
+    return () => promise;
+}
+
+const isIE = window.navigator.userAgent.indexOf("MSIE ") > -1 || window.navigator.userAgent.indexOf("Trident/") > -1; // Remove this line to use Angular Universal
+
+export function loggerCallback(logLevel: LogLevel, message: string) {
+  console.log(message);
+}
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:350]:
```suggestion
}

```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 8]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+export function MSALInterceptorConfigFactory(config: ConfigService): MsalInterceptorConfiguration {
+    const protectedResourceMap = new Map<string, Array<string>>(config.getSettings('interceptor').protectedResourceMap)
+  
+    return {
+      interactionType: config.getSettings('interceptor').interactionType,
+      protectedResourceMap
+    };
+  }
+  
+  export function MSALGuardConfigFactory(config: ConfigService): MsalGuardConfiguration {
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:374]:
```suggestion
export function MSALGuardConfigFactory(config: ConfigService): MsalGuardConfiguration {
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 9]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+    return { 
+      interactionType: config.getSettings('guard').interactionType,
+      authRequest: config.getSettings('guard').authRequest,
+      loginFailedRoute: config.getSettings('guard').loginFailedRoute
+    };
+  }
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:380]:
```suggestion
}
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 10]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+import { MatButtonModule } from '@angular/material/button';
+import { MatToolbarModule } from '@angular/material/toolbar';
+import { MatListModule } from '@angular/material/list';
+
+import { AppRoutingModule } from './app-routing.module';
+import { AppComponent } from './app.component';
+import { HomeComponent } from './home/home.component';
+import { ProfileComponent } from './profile/profile.component';
+
+import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:440]:
```suggestion
import { HttpClientModule } from '@angular/common/http';
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 11]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+import { IPublicClientApplication, PublicClientApplication, InteractionType, BrowserCacheLocation, LogLevel } from '@azure/msal-browser';
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:441]:
```suggestion

```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 12]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+import { MsalGuard, MsalInterceptor, MsalBroadcastService, MsalInterceptorConfiguration, MsalModule, MsalService, MSAL_GUARD_CONFIG, MSAL_INSTANCE, MSAL_INTERCEPTOR_CONFIG, MsalGuardConfiguration, MsalRedirectComponent } from '@azure/msal-angular';
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:442]:
```suggestion
import { MsalRedirectComponent } from '@azure/msal-angular';
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 13]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+import { DetailComponent } from './detail/detail.component';
+import { MsalConfigDynamicModule } from './msal-config-dynamic.module';
+
+
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:446]:
```suggestion
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 14]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+    BrowserModule,
+    BrowserAnimationsModule,
+    AppRoutingModule,
+    MatButtonModule,
+    MatToolbarModule,
+    MatListModule,
+    HttpClientModule,
+    MsalConfigDynamicModule.forRoot('assets/configuration.json')
+  ],
+  providers: [
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:464]:
```suggestion
  providers: [],
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 15]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+  ],
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:465]:
```suggestion
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 16]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+  export function MSALGuardConfigFactory(config: ConfigService): MsalGuardConfiguration {
+    return { 
+      interactionType: config.getSettings('guard').interactionType,
+      authRequest: config.getSettings('guard').authRequest,
+      loginFailedRoute: config.getSettings('guard').loginFailedRoute
+    };
+  }
+
+@NgModule({
+    providers: [
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:383]:
```suggestion
    providers: [],
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 17]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
...
+    ],
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:384]:
```suggestion
```

File: lib/msal-angular/docs/v2-docs/configuration.md [hunk 18]
```diff
@@ -270,3 +271,231 @@ fetch('/assets/configuration.json')
   }
 }
 ```
+
+## Dynamic configurations using Factory Providers and APP_INITIALIZER
+
+To dynamically configure MSAL Angular, you can use the Factory Providers with APP_INITIALIZER
```
Comment (jo-arroyo) [inline lib/msal-angular/docs/v2-docs/configuration.md:277]:
```suggestion
To dynamically configure MSAL Angular, you can use the Factory Providers with APP_INITIALIZER.
```

**Review by jo-arroyo — COMMENTED** [review]
@vinusorout Thanks for your patience. I've tried out your updated configurations, and requested the following changes.

**Comment by vinusorout** [comment]
Hello @jo-arroyo thanks for the suggestions , please let me know if anything left from my side for this PR!

**Review by jo-arroyo — APPROVED** [review]

**Comment by vinusorout** [comment]
Hello @jasonnutter , could you please review these changes! so that we can close this asap.

**Review by jasonnutter — APPROVED** [review]
Please resolve conflicts, thanks!

**Comment by vinusorout** [comment]
Thanks @jasonnutter , I have resolved the conflicts.

**Pull request merged**